### PR TITLE
[processing] QgsProcessingParameterMeshLayer

### PR DESCRIPTION
**Pull request by PeterPetrik**
ability to select mesh layers in the processing algorithms 

some "prototype" algorithms will come shorty in crayfish plugin.

### CODE REVIEW

File: python/plugins/processing/gui/wrappers.py [hunk 1]
```diff
@@ -638,10 +641,15 @@ def _getOptions(self):
                 [QgsProcessingOutputRasterLayer,
                  QgsProcessingOutputMapLayer,
                  QgsProcessingOutputMultipleLayers])
+        elif self.parameterDefinition().layerType() == QgsProcessing.TypeMesh:
+            options = self.dialog.getAvailableValuesOfType(
+                (QgsProcessingParameterMeshLayer, QgsProcessingParameterMultipleLayers),
```
Comment (nyalldawson) [inline python/plugins/processing/gui/wrappers.py]: Should also include QgsProcessingOutputMapLayer

File: src/core/processing/qgsprocessingutils.cpp [hunk 2]
```diff
@@ -75,6 +76,28 @@ QList<QgsVectorLayer *> QgsProcessingUtils::compatibleVectorLayers( QgsProject *
   return layers;
 }
 
+QList<QgsMeshLayer *> QgsProcessingUtils::compatibleMeshLayers( QgsProject *project, bool sort )
+{
+  if ( !project )
+    return QList<QgsMeshLayer *>();
+
+  QList<QgsMeshLayer *> layers;
+  Q_FOREACH ( QgsMeshLayer *l, project->layers<QgsMeshLayer *>() )
```
Comment (nyalldawson) [inline src/core/processing/qgsprocessingutils.cpp]: Q_FOREACH -> range based for

File: src/core/processing/qgsprocessingutils.cpp [hunk 3]
```diff
@@ -85,6 +108,8 @@ QList<QgsMapLayer *> QgsProcessingUtils::compatibleLayers( QgsProject *project,
     layers << rl;
   Q_FOREACH ( QgsVectorLayer *vl, compatibleVectorLayers( project, QList< int >(), false ) )
     layers << vl;
+  Q_FOREACH ( QgsMeshLayer *vl, compatibleMeshLayers( project, false ) )
```
Comment (nyalldawson) [inline src/core/processing/qgsprocessingutils.cpp]: range based for

File: src/core/processing/qgsprocessingutils.h [hunk 4]
```diff
@@ -69,12 +71,28 @@ class CORE_EXPORT QgsProcessingUtils
...
+     * If the \a sort argument is true then the layers will be sorted by their QgsMapLayer::name()
+     * value.
+     *
+     * \see compatibleRasterLayers()
+     * \see compatibleVectorLayers()
+     * \see compatibleLayers()
+     *
+     * \since QGIS 3.6
+     */
+    static QList<QgsMeshLayer *> compatibleMeshLayers( QgsProject *project, bool sort );
```
Comment (nyalldawson) [inline src/core/processing/qgsprocessingutils.h]: Needs unit tests

File: src/core/qgsmaplayerproxymodel.cpp [hunk 5]
```diff
@@ -127,6 +129,7 @@ bool QgsMapLayerProxyModel::filterAcceptsRow( int source_row, const QModelIndex
   // layer type
   if ( ( mFilters.testFlag( RasterLayer ) && layer->type() == QgsMapLayer::RasterLayer ) ||
        ( mFilters.testFlag( VectorLayer ) && layer->type() == QgsMapLayer::VectorLayer ) ||
+       ( mFilters.testFlag( MeshLayer ) && layer->type() == QgsMapLayer::MeshLayer ) ||
```
Comment (nyalldawson) [inline src/core/qgsmaplayerproxymodel.cpp]: Cover in test_qgsmaplayerproxymodel.py

**Review by nyalldawson — COMMENTED** [review]
Looking good! There's some more bits missing here:

- you need an equivalent of QgsProcessingParameterTypeRasterLayer from qgsprocessingparametertypeimpl.h, registered in the QgsProcessingRegistry constructor

- you need to update python/processing/algfactory.py 

- you need to add a your parameter type in processing/tests/GuiTest.py so that it runs against some standard widget tests

**Comment by nyalldawson** [comment]
Oh - question -- is there a need yet for mesh layeroutputs? Or is the output from these algorithms currently vector/raster layers?

**Comment by PeterPetrik** [comment]
@nyalldawson current output of these algorithms is vector/raster layer since we do not have write-API in MDAL at the moment.

File: python/plugins/processing/gui/wrappers.py (hunk 1 above)
Comment (PeterPetrik) [reply python/plugins/processing/gui/wrappers.py]: even there is no mesh write API?

Comment (PeterPetrik) [reply python/plugins/processing/gui/wrappers.py]: @nyalldawson should I add QgsProcessingOutputMapLayer even there is no write support?

Comment (nyalldawson) [reply python/plugins/processing/gui/wrappers.py]: I guess not - we'll just have to remember to add this when mdal inevitably gets write support!

**Pull request merged**